pubtools/pulplib/_impl/model/repository/file.py

```diff
@@ -20,3 +29,66 @@ class FileRepository(Repository):
     )
 
     mutable_urls = attr.ib(default=attr.Factory(lambda: ["PULP_MANIFEST"]), type=list)
+
+    def upload_file(self, file_obj, relative_url=None):
+        """Upload a file to this repository.
+
+        Args:
+            file_obj (str, file object)
+                If it's a string, then it's the path of file to upload.
+                Else, it ought to be a file-like object
+
+            relative_url (str)
+                Path that should be used in remote repository
+
+                If omitted, filename will be used.
+        Returns:
+            Future[:class:`~pubtools.pulplib.Task`]
+                A future which is resolved when import succeeds.
+
+                The future contains the task to import uploaded content
+                to repository
+
+        Raises:
+            DetachedException
+                If this instance is not attached to a Pulp client.
+        """
+        if not self._client:
+            raise DetachedException()
+
+        relative_url = self._get_relative_url(file_obj, relative_url)
+
+        # request upload id and wait for it
+        upload_id = self._client._request_upload().result()["upload_id"]
+
+        upload_complete_f, checksum, size = self._client._do_upload_file(
+            upload_id, self.id, file_obj
+        )
+
+        unit_key = {"name": relative_url, "digest": checksum, "size": size}
+
+        import_complete_f = f_flat_map(
+            upload_complete_f,
+            lambda _: self._client._do_import(self.id, upload_id, "iso", unit_key),
+        )
+
+        f_map(
+            import_complete_f, lambda _: self._client._delete_upload_request(upload_id)
+        )
+
+        return import_complete_f
+
+    def _get_relative_url(self, file_obj, relative_url):
+        is_path = isinstance(file_obj, str)
+        if is_path:
+            if not relative_url:
+                relative_url = file_obj
+            elif relative_url.endswith("/"):
```

Comment:
Can you explain what this is about?  It seems to be a kind of feature:

If you specify "foo/" as relative_url, and the file is named "myfile.txt", then they're joined so the file is uploaded as "foo/myfile.txt" ?

This behavior isn't documented anywhere.  Can you please strip out the behavior, or include it in the documentation for the method?  Looks like we should have some more details on how relative_url is used.

Reply:
Yeah, it's the feature you thought. 
In the previous commit, you mentioned that it should allow users to specify both path and name by passing relative_url, So I implemented like this, make it more flexible. 
Not sure if this is good or not.

Reply:
I think it's a decent feature.  Thinking about if we don't have this feature - the user passing a filename with trailing "/" doesn't really make sense (I don't think Pulp can store/publish a file like that), so we would be leaving the behavior undefined or adding some code to make it crash.

So let's keep it & just make sure it's documented/tested.